packages/react-reconciler/src/ReactFiberCacheComponent.old.js

```diff
@@ -57,6 +62,71 @@ let pooledCache: Cache | null = null;
 // cache from the render that suspended.
 const prevFreshCacheOnStack: StackCursor<Cache | null> = createCursor(null);
 
+// Creates a new empty Cache instance with a ref-count of 0. The caller is responsible
+// for retaining the cache once it is in use (retainCache), and releasing the cache
+// once it is no longer needed (releaseCache).
+let _cacheIndex = 0;
+export function createCache(): Cache {
+  const cache: Cache = {
+    controller: new AbortController(),
+    data: new Map(),
+    refCount: 0,
+  };
+
+  // TODO: remove debugging code
+  if (__DEV__) {
+    cache.key = '' + _cacheIndex++;
+    trace(`createCache #${cache.key ?? ''}`);
+  }
+
+  return cache;
+}
+
+export function retainCache(cache: Cache) {
+  if (__DEV__) {
+    trace(
+      `retainCache #${cache.key ?? ''} ${cache.refCount} -> ${cache.refCount +
+        1}`,
+    );
+  }
+  cache.refCount++;
+}
+
+// Cleanup a cache instance, potentially freeing it if there are no more references
+export function releaseCache(cache: Cache) {
+  if (__DEV__) {
+    trace(
+      `releaseCache #${cache.key ?? ''} ${cache.refCount} -> ${cache.refCount -
+        1}`,
+    );
+  }
+  cache.refCount--;
+  if (__DEV__) {
+    if (cache.refCount < 0) {
+      throw new Error(
```

Comment:
Nit: Let's make this a dev warning. We try not to throw only in dev or only in prod since it diverges program behavior. Not that the two are 100% identical anyway since we do wild stuff like double invoke render functions :D But the other reason is it's less code/overhead in production.